Bklaus/irtrans to tcp and cleanup

klues opened the pull request:
just created this pull-request to have a diff-view which I can cite in my thesis (AsTeRICS-Ergo). But at some time maybe these changes also should be merged.

deinhofer commented:
looks great!!
Please also update the respective IRTrans.htm help file and describe the learning capabilities. Maybe also link to the command specification if necessary.
Then the PR can be merged.

klues commented on ARE/components/actuator.irtrans/src/main/java/eu/asterics/component/actuator/irtrans/IrTransInstance.java:
```diff
@@ -213,116 +243,96 @@ public void receiveEvent(final String data) {
         }
     }
 
-    private void receiveCommand() {
```
question: can we really remove this? it is actually a breaking change. before: irtrans module received commands all the time and sent it to output port. now: only responses to a sent command are sent to output port.
@ChrisVeigl what do you think?

klues commented:
I now updated the code in order to send normal received commands to port "output" and results of sent actions to "outputResult". Therefore we should be compatible with the old version.
I also updated the help file of IrTrans therefore I think we can merge this PR now.

deinhofer commented on ARE/components/actuator.irtrans/src/main/java/eu/asterics/component/actuator/irtrans/IrTransInstance.java:
```diff
@@ -240,42 +247,58 @@ public void receiveData(byte[] data) {
+                    }
+                } catch (Exception e) {
+                    logger.log(Level.INFO, "reading from socket to receive command stopped");
+                }
```
closing the tcpSocketRead should be done in a finally block here, to ensure that it is closed both in case of exception and also after normal interruption.

finally {
     if(tcpSocketRead !=null) tcpSocketRead.close();
}

deinhofer commented on ARE/components/actuator.irtrans/src/main/java/eu/asterics/component/actuator/irtrans/IrTransInstance.java:
```diff
@@ -294,43 +318,43 @@ public void start() {
     public void stop() {
-        closeSocket();
+        closeTcpSocket(tcpSocket);
+        closeTcpSocket(tcpSocketRead);
```
close tcpSocketRead in the reading thread, because here you can't be sure that the reference variable already holds the reference to the socket although it was already opened. This is due to thread caching. To prevent caching you could declare tcpSocketRead as

private volatile Socket tcpSocketRead

nevertheless it's better to close it in the reader thread.

deinhofer commented on ARE/components/actuator.irtrans/src/main/java/eu/asterics/component/actuator/irtrans/IrTransInstance.java:
```diff
@@ -294,43 +318,43 @@ public void start() {
         // string with single space
         byte[] sendBytes = ("A" + cmdString.trim().replaceAll("\\s+", " ") + "\n").getBytes();
-        socket.getInputStream().skip(socket.getInputStream().available());
+        socket.getInputStream().skip(tcpSocket.getInputStream().available());
```
you directly use the member variable tcpSocket, but shouldnt it be the parameter socket?
